File: app/src/main/java/com/alcatraz/admin/project_alcatraz/Auth/AuthViewModel.java
package com.alcatraz.admin.project_alcatraz.Auth;

import android.app.Application;
import android.arch.lifecycle.AndroidViewModel;
import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MediatorLiveData;
import android.arch.lifecycle.ViewModel;
import android.arch.lifecycle.ViewModelProvider;
import android.support.annotation.NonNull;

import com.alcatraz.admin.project_alcatraz.Data.Resource;

import java.util.HashMap;
import java.util.Map;


public class AuthViewModel extends AndroidViewModel{
    private final AuthRepository mRepository;
    private MediatorLiveData<Resource<Map<String, String>>> mData = new MediatorLiveData<>();

    public AuthViewModel(@NonNull Application application) {
        super(application);
        mRepository = AuthRepository.getInstance(application);
    }

    public LiveData<Resource<Map<String, String>>> getObservableData() {
        return mData;
    }

    public void login(String username, String password) {
        Map<String, String> input = new HashMap<>(2);
        input.put("username", username);
        input.put("password", password);
        mData.addSource(mRepository.login(input), mData::setValue);
    }

    public static class Factory extends ViewModelProvider.NewInstanceFactory {
        @NonNull Application mApplication;

        public Factory(@NonNull Application application) {
            mApplication = application;
        }

        @NonNull
        @Override
        public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
            return (T) new AuthViewModel(mApplication);
        }
    }
}

File: app/src/main/java/com/checkin/app/checkin/auth/AuthViewModel.kt
package com.checkin.app.checkin.auth

import android.app.Application
import androidx.lifecycle.MediatorLiveData
import androidx.lifecycle.MutableLiveData
import com.checkin.app.checkin.data.BaseViewModel
import com.checkin.app.checkin.data.resource.Resource
import com.fasterxml.jackson.databind.JsonNode

class AuthViewModel(application: Application) : BaseViewModel(application) {
    private val mRepository: AuthRepository = AuthRepository.getInstance(application)

    private val mErrors = MutableLiveData<JsonNode>()
    private val mAuthResult = createNetworkLiveData<AuthResultModel>()

    var providerIdToken: String? = null

    val authResult: MediatorLiveData<Resource<AuthResultModel>> = mAuthResult

    override fun updateResults() {}

    fun authenticate(idToken: String) {
        val reqData = AuthRequestModel(idToken, providerIdToken ?: "")
        mAuthResult.addSource(mRepository.authenticate(reqData), mAuthResult::setValue)
    }

    companion object {
        private val TAG = AuthViewModel::class.java.simpleName
    }
}